### PR TITLE
[WIP] Implement RFC 1260

### PR DESCRIPTION
**Pull request by F001**
This is my in-progress work to implement RFC 1260: main_reexport. 

Tracking issue: https://github.com/rust-lang/rust/issues/28937

There is a known problem that, I have no idea how to get the foreign function's signature if it is imported from another crate. Thus, the type check for this case is not done yet. I'd appreciate it if someone can provide some helps.

### CODE REVIEW

**Comment by petrochenkov** [comment]
@rust-lang/lang 
This is one of the worst accepted RFCs that I've seen in terms of cost vs benefit.
It adds a weird corner case to the language and compiler while all it does is changing `fn main() { my_fn() }` to `use my_fn as main;`.
We should rather unaccept in than implement.

**Comment by jonas-schievink** [comment]
This does seem like it adds a lot of corner cases in the compiler. For example, we already have error codes for doing `use A as X; use B as X;` as well as `fn main<T>() {}`. Those could be unified with the ones added by this PR.

Similarly, the distinction of "imported main" vs. "regular fn main" could be removed. It seems to me that `main` is currently rather special and from my naive model of how this part of rustc works it might be desirable to instead do the equivalent of "the item named `main` is the program's main function" and go through the regular resolve steps that will automatically handle `use`.

I disagree that this is a special case in the language itself, though. If we adopt the "the item named `main` is the program's main function" model this would be entirely consistent.

**Comment by eddyb** [comment]
One of the annoying bits is that `main` is type-checked specially - if we were to do this *entirely separately*, based on the type of the item that we've chosen as our `main` (which we'd refer to by `DefId` alone), then it should "just work", even cross-crate. I think we already use just the `DefId` to generate the function pointer to pass to `libstd`, so that part should be fine.

**Comment by F001** [comment]
@jonas-schievink Good suggestion. I'll try to re-use existing code as much as possible.

@eddyb I have stored the `DefId` of imported `main` in `config::EntryFnType::EntryImported`. For type check, I was trying to call existing function:  https://github.com/rust-lang/rust/blob/5d0631a6438cf30cac236b7176371663e35c8d07/src/librustc_typeck/lib.rs#L178-L180

This function requires a `ast::NodeId` as the parameter. I tried to get the `NodeId` from `DefId` as below:

```rust
                if let Some(node_id) = tcx.hir.as_local_node_id(*def_id) {
                    check_main_fn_ty(tcx, node_id, *def_id, *sp);
                } else {
                    span_bug!(*sp, "NodeId of `main` is not found");
                }
```
Then I encountered the problem: `as_local_node_id` returns `None` in cross-crate case.

This is my first pull request in the rust compiler, and I must have made some mistakes somewhere due to lack of knowledge of rustc internal.

**Comment by eddyb** [comment]
@F001 You should go the other way around and change the current code to use `DefId`s more.
In fact, you can do all of that as a refactor PR without introducing the new feature, just to make the PR that *does* introduce the feature much smaller.

**Comment by F001** [comment]
@eddyb Thanks. I'll close this PR later, and split it into 2 PRs.

But in order to refactor `check_main_fn_ty`, I still have to address the original question first. 

The main problem is that this function uses `hir::map::Map::find(&self, NodeId) -> Option<Node<'hir>>`  to get the actual `hir::map::Node`, and then the `generics` check can be done.

When I was trying to use `DefId` as the argument of the function `check_main_fn_ty`, the only way I know is calling `tcx.hir.as_local_node_id(def_id)` first and then using `tcx.hir.find(node_id)`. However, this doesn't work in cross-crate case.

In other words, I can't find a way to do the refactoring to use `DefId` replace `NodeId` in `check_main_fn_ty`.

Finally, the question still exists. How can I get the foreign function's signature from `DefId` ?

**Comment by eddyb** [comment]
You'll have to work with `tcx.generics_of(def_id)` and `tcx.type_of(def_id)`. May need to also use `tcx.describe_def(def_id)` to make sure it's a function to begin with?

**Comment by F001** [comment]
@eddyb Thanks for the information. I'll have a try later. This PR can be closed now.

**Comment by petrochenkov** [comment]
@jonas-schievink 
>If we adopt the "the item named main is the program's main function" model this would be entirely consistent.

I agree that such "resolution-based" formulation (not even mentioning reexports) would be better.
- Resolve path `crate::main` (with all idents having non-macro contexts) using usual name resolution and whatever is found is our main.
This resolution can be done during lowering to HIR, so we can just keep main in HIR as an `Option<DefId>` field for later type checking.
